Update XR interface and remove XRDevice

toji opened the pull request:
Fixes #55. Also changes the syntax of the WebGL context compatibility
methods because they required an XRDevice to function properly.

This required more restructuring than I expected (but it should have been obvious in retrospect.) It does introduce one polyfill-specific error, in which developers are required to have confirmed their desired immersive mode is supported before calling requestSession. This is due to the WebVR implementation having a two-step async process in the equivalent area instead of a one-step, which may cause the user gesture token to get lost if the developer hasn't taken some other asynchronous action prior to calling requestSession. Not happy with that, but at least it reinforces best practices rather than being something totally arbitrary. 

Note that this change leaves inline content non-functional because the changes listed in #53 are required to get that portion of the API back up to spec, and in the meantime there's no way to set the output context.

jsantell commented:
Thanks @toji! Just taking a look at this now, juggling the user gesture is certainly non-trivial :sob:

toji commented:
Sorry, I've been working on updating the tests for this but my laptop just flipped out and needed to be wiped, so I lost my progress on that part. :( I'll see what I can do to bring it back up again, but I'll warn you now that the tests changes are pretty extensive.

jsantell commented on src/polyfill-globals.js:
```diff
@@ -15,38 +15,32 @@
  */
-export const polyfillSetCompatibleXRDevice = Context => {
-  if (typeof Context.prototype.setCompatibleXRDevice === 'function') {
+export const polyfillMakeXRCompatible = Context => {
```
If we have challenges with this area, I'd be ok with removing it; idea is to be able to test with the polyfill and it can catch early errors like not making the context compatible even if you don't have access to a working WebXR browser (although, less of an issue now I think thanks to GVR+Chrome?) -- or can handle in a follow up

toji commented on src/api/XR.js:
```diff
@@ -17,26 +17,101 @@ import EventTarget from '../lib/EventTarget';
+   * @return {Promise<null>}
+   */
+  async supportsSession(mode) {
+    if (!mode in XRSessionModes) {
```
Derp. This is the first of many "Wait, I know I saw this working when I wrote it, but... how?" issues that came up in this review. Thanks for your patience!

toji commented on src/api/XR.js:
```diff
@@ -17,26 +17,101 @@ import EventTarget from '../lib/EventTarget';
     super();
     this[PRIVATE] = {
-      device,
+      polyfill,
```
I agree! (Though this made the change a fair amount larger.) It's now XRDevice everywhere (except the tests)

toji commented on src/api/XR.js:
```diff
@@ -17,26 +17,101 @@ import EventTarget from '../lib/EventTarget';
+        // Because requesting immersive modes requires a user gesture, we can't
+        // wait for a promise to resolve before making the real session request.
+        // For that reason, we'll throw a polyfill-specific error here.
+        throw new Error(POLYFILL_REQUEST_SESSION_ERROR);
```
Just to confirm: When we talked on the phone last week I seem to recall that the decision we made here was that the logic was OK as-is? Happy to refactor the code if that's not what you remember.

toji commented on src/api/XRSession.js:
```diff
@@ -25,48 +25,23 @@ import XRFrameOfReference, {
-  constructor(polyfill, device, sessionOptions, id) {
-    sessionOptions = Object.assign({}, XRSessionCreationOptions, sessionOptions);
-
+  constructor(polyfill, device, mode, id) {
```
Fixed along with the `XRDevice` change.

toji commented on src/api/XRWebGLLayer.js:
```diff
@@ -74,26 +75,11 @@ export default class XRWebGLLayer extends XRLayer {
-  /**
-   * @return {boolean}
-   */
-  get alpha() { return this[PRIVATE].config.alpha; }
```
These were actually removed from the spec, and it was an area where since I was messing around with the file anyway I figured I'd just go ahead and make the signature match.

toji commented on src/devices.js:
```diff
@@ -13,28 +13,12 @@
  * limitations under the License.
  */
 
-import XRDevice from './api/XRDevice';
```
👍

toji commented on src/devices.js:
```diff
@@ -54,28 +38,21 @@ const getVRDisplay = async function (global) {
-      return new XRDevice(device);
+    xr = await getVRDisplay(global);
+    if (xr) {
+      return xr;
```
Actually it's the reverse. Since the promise is being resolved within the XR object now, and we want that to be immediately available, the line futher down (wrapping the Cardboard device in `new XR()`) is wrong and this one was OK. The construction of the `XR` interface is happening in WebXRPolyfill.js

toji commented on src/devices/WebVRDevice.js:
```diff
@@ -52,9 +48,10 @@ const PRIMARY_BUTTON_MAP = {
+  constructor(mode, polyfillOptions={}) {
+    this.mode = mode;
+    this.outputContext = null;
+    this.immersive = mode == 'immersive-vr' || 'immersive-ar';
```
😓 Seriously, me? seriously?

toji commented:
Made a bunch of updates in response to your feedback and our call. Sorry it's been taking so long to push through. I haven't had a chance to test this everywhere I'd like just yet, but it should be much closer to a mergeable state now.

jsantell commented on build/webxr-polyfill.module.js:
```diff
@@ -141,19 +141,68 @@ class EventTarget {
 }
```
We can save checking in the built files until we're ready to merge back to master

toji commented on src/api/XR.js:
```diff
@@ -17,26 +17,103 @@ import EventTarget from '../lib/EventTarget';
+   * @return {Promise<null>}
+   */
+  async supportsSession(mode) {
+    if (XRSessionModes.indexOf(mode) == -1) {
```
SGTM!

toji commented:
Okay! Addressed your latest feedback, re-targeted this PR at the webxr-2019 branch, squashed the commits to keep the branch history a bit cleaner, and merged! Thank you for the reviews and sorry for the slow dev cycle on this one.

jsantell commented:
:tada:  thanks @toji!
